### PR TITLE
[REVIEW] Add GroupBy.groups

### LINKED ISSUE #5624: Groupby df columns and get groups keys[BUG]
**Describe the bug**
cudf.DataFrame.groupby(by=['column1', 'column2']).groups.keys() produces the following error:
ValueError: Grouper and object must have same length

I'm able to run DataFrame.groupby(['column1', 'column2']).groups.keys() without any problem.

**Environment**
Colab

**Environment details**
# Install RAPIDS
!git clone https://github.com/rapidsai/rapidsai-csp-utils.git
!bash rapidsai-csp-utils/colab/rapids-colab.sh stable

import sys, os

dist_package_index = sys.path.index('/usr/local/lib/python3.6/dist-packages')
sys.path = sys.path[:dist_package_index] + ['/usr/local/lib/python3.6/site-packages'] + sys.path[dist_package_index:]
sys.path
exec(open('rapidsai-csp-utils/colab/update_modules.py').read(), globals())

**Additional context**
'column1' and 'column2' dtypes are both int64

[cudf.DataFrame.groupby error.docx](https://github.com/rapidsai/cudf/files/4865627/cudf.DataFrame.groupby.error.docx)

### PR DESCRIPTION
**Pull request by shwina**
Closes #5624

### CODE REVIEW

File: python/cudf/cudf/core/groupby/groupby.py [hunk 1]
```diff
@@ -64,10 +64,20 @@ def __getattribute__(self, key):
...
             yield name, grouped_values[offsets[i] : offsets[i + 1]]
 
+    @cached_property
+    def groups(self):
+        """
+        Returns a dictionary mapping group keys to row labels.
+        """
+        group_names, offsets, _, _ = self._grouped()
+        return dict(
+            zip(group_names.to_pandas(), self.obj.index._split(offsets[1:-1]))
```
Comment (shwina) [inline python/cudf/cudf/core/groupby/groupby.py:78]: Is this what we should return?

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: python/cudf/cudf/core/groupby/groupby.py (hunk 1 above)
Comment (shwina) [reply python/cudf/cudf/core/groupby/groupby.py:78]: cc: @kkraus14

File: python/cudf/cudf/core/frame.py [hunk 2]
```diff
@@ -815,6 +815,14 @@ def mask(self, cond, other=None, inplace=False):
 
         return self.where(cond=~cond, other=other, inplace=inplace)
 
+    def _split(self, splits, keep_index=True):
+        result = libcudf.copying.table_split(
+            self, splits, keep_index=keep_index
+        )
+
+        result = [self.__class__._from_table(tbl) for tbl in result]
```
Comment (jrhemstad) [inline python/cudf/cudf/core/frame.py:823]: Is this making a deep copy of all of the splits?

Comment (shwina) [reply python/cudf/cudf/core/frame.py:823]: Nope, it does the following:

1. It wraps the table (views) returned by libcudf in an object more familiar to our users (Index/DataFrame/Series/etc.,). This is the `self__class__._from_table(tbl)` bit. This doesn't copy any GPU data.

2. Creates a list of those. This doesn't copy anything either.

File: python/cudf/cudf/core/groupby/groupby.py (hunk 1 above)
Comment (kkraus14) [reply python/cudf/cudf/core/groupby/groupby.py:79]: Yes this is what we should return, but I think we want to add a heuristic to this and other iteration based methods that over say 100 or 1000 groups there's going to be significant performance degradation.

Comment (shwina) [reply python/cudf/cudf/core/groupby/groupby.py:79]: 👍

File: python/cudf/cudf/core/groupby/groupby.py [hunk 3]
```diff
@@ -76,6 +80,13 @@ def groups(self):
         """
         group_names, offsets, _, grouped_values = self._grouped()
         grouped_index = grouped_values.index
+
+        if len(group_names) > self._MAX_GROUPS_BEFORE_WARN:
+            warnings.warn(
+                f"GroupBy.groups() expected to be slow"
+                f" for large number of groups ({len(group_names)})"
```
Comment (kkraus14) [inline python/cudf/cudf/core/groupby/groupby.py:87]: We should make this more clear what the number is: because right now it would output:
```
GroupBy.groups() expected to be slow for large number of groups 1000
```

Maybe reword to:
```
GroupBy.groups() performance scales poorly with a large number of groups: got 1000 groups
```?

Comment (shwina) [reply python/cudf/cudf/core/groupby/groupby.py]: Yeah I definitely like warning message better. Changed it to that
